File: samples/c/ftl/tibthrurecv.c
/*
 * Copyright (c) 2010-2018 TIBCO Software Inc.
 * Licensed under a BSD-style license. Refer to [LICENSE]
 * For more information, please contact:
 * TIBCO Software Inc., Palo Alto, California, USA
 */

/*
 * This is a sample C responder program which is used with the tibthrusend
 * program to measure basic throughput between the publisher and subscriber.
 */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#include "tib/ftl.h"

#include "tibaux.h"

#define DEFAULT_URL "http://localhost:8080"

const char*     realmServer             = DEFAULT_URL;
const char*     appName                 = "tiblatrecv";
const char*     endpointName            = "tiblatrecv-sendendpoint";
const char*     recvEndpointName        = "tiblatrecv-recvendpoint";
const char*     durableName             = NULL;
const char*     user                    = "guest";
const char*     password                = "guest-pw";
const char      *trcLevel               = NULL;
tibbool_t       inline_mode             = tibfalse;
tibbool_t       doEcho                  = tibtrue;
tibbool_t       doAdvisory              = tibtrue;
const char      *identifier             = NULL;
tibbool_t       trustAll                = tibfalse;
const char      *trustFile              = NULL;
const char      *clientLabel            = NULL;

tibPublisher    pub                     = NULL;
tibFieldRef     dataRef                 = NULL;
tibMessage      reportMsg               = NULL;
volatile tibbool_t  running             = tibtrue;

tibint64_t      start;
tibint64_t      count;
tibint64_t      bytes;
tibAux_StatRecord    blockSizeStats          = {0, 0, 0, 0.0, 0.0};


void
usage(int argc, char** argv)
{
    printf("Usage: %s [options] url\n", argv[0]);
    printf("Default url is "DEFAULT_URL"\n");
    printf("   -a,  --application <name>        Application name\n"
           "        --client-label <label>      Set client label\n"
           "   -d,  --durableName <name>        Durable name for receive side.\n"
           "   -et, --tx-endpoint <name>        Transmit endpoint name\n"
           "   -er, --rx-endpoint <name>        Receive endpoint name\n"
           "   -h,  --help                      Print this help.\n"
           "   -i,  --inline                    Put EventQueue in INLINE mode.\n"
           "   -id, --identifier                Choose instance, eg, \"rdma\"\n"
           "   -n,  --no-echo                   Do not echo messages.\n"
           "   -na, --no-advisory               Do not create a subscriber for dataloss advisories\n"
           "   -p,  --password <string>         Password for realm server.\n"
           "   -t,  --trace <level>             Set trace level\n"
           "           where <level> can be:\n"
           "                 off, severe, warn, info, debug, verbose\n"
           "   -u,  --user <string>             User for realm server.\n"
           "        --trustall\n"
           "        --trustfile <file>\n");

    exit(1);
}

void
parseArgs(int argc, char** argv)
{
    int     i;

    tibAux_InitGetArgs(argc, argv, usage);

    for (i = 1;  i < argc;  i++) {
        if (tibAux_GetString(&i, "--application", "-a", &appName))
            continue;
        if (tibAux_GetString(&i, "--client-label", "--client-label", &clientLabel))
            continue;
        if (tibAux_GetString(&i, "--tx-endpoint", "-et", &endpointName))
            continue;
        if (tibAux_GetString(&i, "--rx-endpoint", "-er", &recvEndpointName))
            continue;
        if (tibAux_GetFlag(&i, "--help", "-h"))
           usage(argc, argv);
        if (tibAux_GetString(&i, "--identifier", "-id", &identifier))
            continue;
        if (tibAux_GetString(&i, "--password", "-p", &password))
            continue;
        if (tibAux_GetString(&i, "--trace", "-t", &trcLevel))
            continue;
        if (tibAux_GetString(&i, "--user", "-u", &user))
            continue;
        if (tibAux_GetString(&i, "--durableName", "-d", &durableName))
            continue;
        if (tibAux_GetFlag(&i, "--inline", "-i")) {
            inline_mode = tibtrue;
            continue;
        }
        if (tibAux_GetFlag(&i, "--no-echo", "-n")) {
            doEcho = tibfalse;
            continue;
        }
        if (tibAux_GetFlag(&i, "--no-advisory", "-na")) {
            doAdvisory = tibfalse;
            continue;
        }
        if (tibAux_GetFlag(&i, "--trustall", "--trustall"))
        {
            if (trustFile)
            {
                printf("cannot specify both --trustall and --trustfile\n");
                usage(argc, argv);
            }
            trustAll = tibtrue;
            continue;
        }
        if (tibAux_GetString(&i, "--trustfile", "--trustfile", &trustFile))
        {
            if (trustAll)
            {
                printf("cannot specify both --trustall and --trustfile\n");
                usage(argc, argv);
            }
            continue;
        }
        if (strncmp(argv[i], "-", 1)==0)
        {
            printf("invalid option: %s\n", argv[i]);
            usage(argc, argv);
        }
        realmServer = argv[i];
    }

    printf("Invoked as: ");
    for (i = 0;  i < argc;  i++)
        printf("%s ", argv[i]);
    printf("\n");
}

void
getResult(char* outbuf, size_t outsize, tibint64_t stop)
{
    double  interval = (stop-start) * tibAux_GetTimerScale();
    tibAux_StringCat(outbuf,outsize,"Received ");
    tibAux_StringCatNumber(outbuf,outsize,(double)count);
    tibAux_StringCat(outbuf,outsize," messages in ");
    tibAux_StringCatNumber(outbuf,outsize,interval);
    tibAux_StringCat(outbuf,outsize," seconds. (");
    tibAux_StringCatNumber(outbuf,outsize,count/interval);
    tibAux_StringCat(outbuf,outsize," messages per second)\n");

    tibAux_StringCat(outbuf,outsize,"Received ");
    tibAux_StringCatNumber(outbuf,outsize,(double)bytes);
    tibAux_StringCat(outbuf,outsize," bytes in ");
    tibAux_StringCatNumber(outbuf,outsize,interval);
    tibAux_StringCat(outbuf,outsize," seconds. (");
    tibAux_StringCatNumber(outbuf,outsize,bytes/interval);
    tibAux_StringCat(outbuf,outsize," bytes per second)\n");
    
    tibAux_StringCat(outbuf,outsize,"Messages per callback: ");
    tibAux_StringCatStats(outbuf,outsize,&blockSizeStats, 1.0);
    tibAux_StringCat(outbuf,outsize,"\n");
}

void
onMsg(
    tibEx                       ex,
    tibEventQueue               msgQueue,
    tibint32_t                  msgNum,
    tibMessage                  *msgs,
    void                        **closures)
{
    int                     i;
    const void*             buf;
    tibint32_t              bufsize;
    tibint32_t              n = msgNum;
    char                    reportBuf[512];

    *reportBuf = '\0';
    
    for (i = 0;  i < msgNum;  i++)
    {
        buf = tibMessage_GetOpaqueByRef(ex, msgs[i], dataRef, &bufsize);
        if (tibEx_GetErrorCode(ex) == TIB_OK)    // There was no error getting the buffer
        {
            switch (*(tibint64_t*)buf)
            {
                case 0:
                    count++;
                    bytes += bufsize;
                    break;
                case -1:
                    start = tibAux_GetTime();
                    count = 0;
                    bytes = 0;
                    blockSizeStats.N = 0;
                    n--;
                    break;
                case -2:
                    getResult(reportBuf, sizeof(reportBuf), tibAux_GetTime());
                    printf("%s\n",reportBuf); fflush(stdout);
                    n--;
                    break;
                case -3:
                    getResult(reportBuf, sizeof(reportBuf), tibAux_GetTime());
                    printf("%s\n",reportBuf); fflush(stdout);

                    //Format and send back the report to the sender
                    tibMessage_SetString(ex, reportMsg, "receiverReport", reportBuf);
                    tibPublisher_Send(ex, pub, reportMsg);

                    *reportBuf = '\0';
                    n--;
                    break;
                case -4:
                    if (doEcho) // Send pong for initial ping.
                        tibPublisher_Send(ex, pub, msgs[i]);
                    n--;
                    break;
                default:
                    if (doEcho)
                        tibPublisher_Send(ex, pub, msgs[i]);
                    count++;
                    bytes += bufsize;
            }
        }
    }
    
    if (tibEx_GetErrorCode(ex) != TIB_OK)
    {
        tibAux_PrintException(ex);
        running = tibfalse;
    }
    
    if (n)
        tibAux_StatUpdate(&blockSizeStats, n);
}

void
onAdvisory(
    tibEx               ex,
    tibEventQueue       msgQueue,
    tibint32_t          msgNum,
    tibMessage          *msgs,
    void                **closures)
{
    tibint32_t          i;

    for (i = 0; i < msgNum; i++)
    {
        printf("received advisory\n");
        tibAux_PrintAdvisory(ex, msgs[i]);
    }

    // any exception that happens in the callback does not
    // propagate to the caller of the tibEventQueue_Dispatch
    if (tibEx_GetErrorCode(ex) != TIB_OK)
        tibAux_PrintException(ex);
}

int
main(int argc, char** argv)
{
    tibRealm                realm   = NULL;
    tibEventQueue           queue   = NULL;
    tibProperties           props   = NULL;
    tibSubscriber           subs    = NULL;
    tibSubscriber           adv     = NULL;
    tibContentMatcher       am      = NULL;
    char                    str[256];
    tibEx                   ex      = tibEx_Create();

    printf("#\n# %s\n#\n# %s\n#\n", argv[0], tib_Version());

    parseArgs(argc, argv);

    tibAux_CalibrateTimer();

    tib_Open(ex, TIB_COMPATIBILITY_VERSION);
    CHECK(ex);

    // Set global trace to specified level
    if (trcLevel != NULL)
    {
        tib_SetLogLevel(ex, trcLevel);
        CHECK(ex);
    }

    // Get a single realm per process
    props = tibProperties_Create(ex);
    tibProperties_SetString(ex, props,
                            TIB_REALM_PROPERTY_STRING_USERNAME, user);
    tibProperties_SetString(ex, props,
                            TIB_REALM_PROPERTY_STRING_USERPASSWORD, password);
    if (identifier)
        tibProperties_SetString(ex, props,
                                TIB_REALM_PROPERTY_STRING_APPINSTANCE_IDENTIFIER, identifier);
    if (trustAll)
        tibProperties_SetLong(ex, props, TIB_REALM_PROPERTY_LONG_TRUST_TYPE,
                              TIB_REALM_HTTPS_CONNECTION_TRUST_EVERYONE);
    if (trustFile)
    {
        tibProperties_SetLong(ex, props, TIB_REALM_PROPERTY_LONG_TRUST_TYPE,
                TIB_REALM_HTTPS_CONNECTION_USE_SPECIFIED_TRUST_FILE);
        tibProperties_SetString(ex, props, TIB_REALM_PROPERTY_STRING_TRUST_FILE, trustFile);
    }

    // set the client label
    if (clientLabel != NULL)
        tibProperties_SetString(ex, props, TIB_REALM_PROPERTY_STRING_CLIENT_LABEL, clientLabel);
    else
        tibProperties_SetString(ex, props, TIB_REALM_PROPERTY_STRING_CLIENT_LABEL, argv[0]);

    realm = tibRealm_Connect(ex, realmServer, appName, props);
    tibProperties_Destroy(ex, props);
    CHECK(ex);
    
    dataRef = tibFieldRef_Create(ex, TIB_BUILTIN_MSG_FMT_OPAQUE_FIELDNAME);
    reportMsg = tibMessage_Create(ex, realm, "ReportMsgFmt");
    CHECK(ex);

    props = tibProperties_Create(ex);
    tibProperties_SetBoolean(ex, props, TIB_EVENTQUEUE_PROPERTY_BOOL_INLINE_MODE, inline_mode);
    CHECK(ex);
    queue = tibEventQueue_Create(ex, realm, props);
    tibProperties_Destroy(ex, props);
    CHECK(ex);

    if (doAdvisory)
    {
        // Create a data loss advisory matcher
        tibAux_snprintf(str, sizeof(str), "{\"%s\":\"%s\"}", 
                      TIB_ADVISORY_FIELD_NAME, TIB_ADVISORY_NAME_DATALOSS);
        am = tibContentMatcher_Create(ex, realm, str);
        // Create an advisory subscriber
        adv = tibSubscriber_Create(ex, realm, TIB_ADVISORY_ENDPOINT_NAME, am, NULL);
        tibEventQueue_AddSubscriber(ex, queue, adv, onAdvisory, NULL);
    }

    if (durableName && *durableName)
    {
        props = tibProperties_Create(ex);
        tibProperties_SetString(ex, props, TIB_SUBSCRIBER_PROPERTY_STRING_DURABLE_NAME, durableName);
    }
    else
    {
        props = NULL;
    }

    subs = tibSubscriber_Create(ex, realm, recvEndpointName, NULL, props);
    tibProperties_Destroy(ex, props);
    CHECK(ex);
    tibEventQueue_AddSubscriber(ex, queue, subs, onMsg, NULL);
    CHECK(ex);

    pub = tibPublisher_Create(ex, realm, endpointName, NULL);
    CHECK(ex);
    
    while (!tibEx_GetErrorCode(ex) && running)
        tibEventQueue_Dispatch(ex, queue, TIB_TIMEOUT_WAIT_FOREVER);
    CHECK(ex);

    tibPublisher_Close(ex, pub);
    if (queue && subs)
        tibEventQueue_RemoveSubscriber(ex, queue, subs, NULL);
    if (queue && adv)
        tibEventQueue_RemoveSubscriber(ex, queue, adv, NULL);
    tibSubscriber_Close(ex, subs);
    tibSubscriber_Close(ex, adv);
    tibContentMatcher_Destroy(ex, am);
    tibEventQueue_Destroy(ex, queue, NULL);
    tibFieldRef_Destroy(ex, dataRef);
    tibMessage_Destroy(ex, reportMsg);
    tibRealm_Close(ex, realm);
    tib_Close(ex);
    tibEx_Destroy(ex);

    return 0;
}

File: samples/java/ftl/TibThruRecv.java
/*
 * Copyright (c) 2013-2017 TIBCO Software Inc.
 * Licensed under a BSD-style license. Refer to [LICENSE]
 * For more information, please contact:
 * TIBCO Software Inc., Palo Alto, California, USA
 */

/*
 * This is a sample java responder program which is used with the TibThruSend
 * program to measure basic throughput between the publisher and subscriber.
 */

package com.tibco.ftl.samples;

import java.text.*;
import java.util.*;

import com.tibco.ftl.*;

public class TibThruRecv
{
    String   appName        = "tiblatrecv";
    String   realmServer    = null;
    String   defaultUrl     = "http://localhost:8080";
    String   defaultSecureUrl = "https://localhost:8080";
    String   endpointName   = "tiblatrecv-sendendpoint";
    String   recvEndpointName = "tiblatrecv-recvendpoint";
    String   durableName    = null;
    String   trcLevel       = null;
    String   user           = "guest";
    String   password       = "guest-pw";
    String   identifier     = null;
    boolean  inline_mode    = false;
    boolean  doEcho         = true;
    boolean  doAdvisory     = true;
    boolean  trustAll       = false;
    String   trustFile      = null;
    String   clientLabel    = null;

    Publisher pub           = null;
    boolean   running       = true;
    MessageFieldRef dataRef = null;
    Message reportMessage   = null;
    long    start;
    long    count;
    long    bytes;
    TibAuxStatRecord blockSizeStats = new TibAuxStatRecord();

    String   usageString    =
        "TibThruRecv [options] url\n" +
        "Default url is " + defaultUrl + "\n" +
        "  where options can be:\n" +
        "    -a,  --application <name>          Application name\n" +
        "    -cl, --client-label <string>       Set client label\n" +
        "    -d,  --durableName <string>        Durable name for receive side\n" +
        "    -et, --tx-endpoint <name>          Transmit endpoint name\n" +
        "    -er, --rx-endpoint <name>          Receive endpoint name\n" +
        "    -h,  --help                        Print this help\n" +
        "    -id, --identifier <string>         Choose instance, eg, \"rdma\"\n" +
        "    -p,  --password <string>           Password for realm server\n" +
        "    -n,  --no-echo                     Do not echo messages.\n" +
        "    -na, --no-advisory                 Do not create a subscriber for dataloss advisories\n" +
        "    -i,  --inline                      Put EventQueue in INLINE mode.\n" +
        "    -t,  --trace <level>               Set trace level\n" +
        "            where <level> can be:\n" +
        "                  off, severe, warn, info, debug, verbose\n" +
        "        --trustall\n" +
        "        --trustfile <file>\n" +
        "    -u,  --user <string>               User for realm server\n";


    public TibThruRecv(String[] args)
    {
        System.out.printf("#\n# %s\n#\n# %s\n#\n",
                          this.getClass().getName(),
                          FTL.getVersionInformation());        
        
        parseArgs(args);
    }
    
    public void usage()
    {
        System.out.printf(usageString);
        System.exit(0);
    }

    public void parseArgs(String args[])
    {
        int i           = 0;
        int argc        = args.length;
        String s        = null;
        TibAux aux      = new TibAux(args, usageString);

        for (i = 0; i < argc; i++)
        {
            s = aux.getString(i, "--application", "-a");
            if (s != null) {
                appName = s;
                i++;
                continue;
            }
            s = aux.getString(i, "--client-label", "-cl");
            if (s != null) {
                clientLabel = s;
                i++;
                continue;
            }
            s = aux.getString(i, "--tx-endpoint", "-et");
            if (s != null) {
                endpointName = s;
                i++;
                continue;
            }
            s = aux.getString(i, "--rx-endpoint", "-er");
            if (s != null) {
                recvEndpointName = s;
                i++;
                continue;
            }
            if (aux.getFlag(i, "--help", "-h")) {
                usage();
            }
            s = aux.getString(i, "--identifier", "-id");
            if (s != null) {
                identifier = s;
                i++;
                continue;
            }
            s = aux.getString(i, "--password", "-p");
            if (s != null) {
                password = s;
                i++;
                continue;
            }
            s = aux.getString(i, "--durable", "-d");
            if (s != null) {
                durableName = s;
                i++;
                continue;
            }
            s = aux.getString(i, "--trace", "-t");
            if (s != null) {
                trcLevel = s;
                i++;
                continue;
            }
            if (aux.getFlag(i, "--trustall", "--trustall")) {
                if (trustFile != null)
                {
                    System.out.println("cannot specify both --trustall and --trustfile");
                    usage();
                }
                trustAll = true;
                continue;
            }
            s = aux.getString(i, "--trustfile", "--trustfile");
            if (s != null)
            {
                trustFile = s;
                if (trustAll)
                {
                    System.out.println("cannot specify both --trustall and --trustfile");
                    usage();
                }
                i++;
                continue;
            }
            s = aux.getString(i, "--user", "-u");
            if (s != null) {
                user = s;
                i++;
                continue;
            }
            if (aux.getFlag(i, "--inline", "-i")) {
                inline_mode = true;
                continue;
            }
            if (aux.getFlag(i, "--no-echo", "-n")) {
                doEcho = false;
                continue;
            }
            if (aux.getFlag(i, "--no-advsiory", "-na")) {
                doAdvisory = false;
                continue;
            }
            if (args[i].startsWith("-", 0)) {
                System.out.println("invalid option: "+args[i]);
                usage();
            }
            realmServer = args[i];
        }

        System.out.print("Invoked as: TibThruRecv ");
        for (i = 0; i < argc; i++)
            System.out.print(args[i]+" ");
        System.out.println("");
    }

    public String getResult(long stop)
    {
        StringBuilder result = new StringBuilder();
        NumberFormat nf = new DecimalFormat(" ##0.00E00 ");
        double interval = (stop-start) * 1.0e-9;
        
        result.append("Received ");
        result.append(nf.format((double)count));
        result.append(" messages in ");
        result.append(nf.format(interval));
        result.append(" seconds. (");
        result.append(nf.format(count/interval));
        result.append(" messages per second)\n");

        result.append("Received ");
        result.append(nf.format((double)bytes));
        result.append(" bytes in ");
        result.append(nf.format(interval));
        result.append(" seconds. (");
        result.append(nf.format(bytes/interval));
        result.append(" bytes per second)\n");
        
        result.append("Messages per callback: ");
        result.append(blockSizeStats.toString(1.0));
        result.append("\n");
        
        return result.toString();
    }

    //convert bytes to a long - java order - requires at least 8 bytes
    public long convertToLong(byte[] bytes)
    {
        long retVal = 0;
        long i;
        int shift;
        int index;

        for(index = 0, shift = 0; index < 8; index++)
        {
            i = (long) bytes[index];
            if(i < 0) i = -(i ^ 0xff) - 1;

            retVal += (i << shift);
            shift += 8;
        }
        
        return retVal;
    }
    
    public void onMsg(List<Message> messages, EventQueue eventQueue)
    {
        int n = messages.size();//block count
        String result = null;
        
        try
        {
            for(int i = 0,msgNum = messages.size();  i < msgNum;  i++)
            {
                byte[] buf = messages.get(i).getOpaque(dataRef);
                int flag = (int) convertToLong(buf);
                switch(flag)
                {
                    case 0:
                        count++;
                        bytes += buf.length;
                        break;
                    case -1:
                        start = System.nanoTime();
                        count = 0;
                        bytes = 0;
                        blockSizeStats.setN(0);
                        n--;
                        break;
                    case -2:
                        result = getResult(System.nanoTime());
                        System.out.printf("%s\n",result); System.out.flush();
                        n--;
                        break;
                    case -3:
                        result = getResult(System.nanoTime());
                        System.out.printf("%s\n",result); System.out.flush();
    
                        //Format and send back the report to the sender
                        reportMessage.setString("receiverReport", result);
                        pub.send(reportMessage);
                        n--;
                        break;
                    case -4:
                        if (doEcho) // Send pong for initial ping.
                            pub.send(messages.get(i));
                        n--;
                        break;
                    default:
                        if (doEcho)
                            pub.send(messages.get(i));
                        count++;
                        bytes += buf.length;
                }
            }
        }
        catch(Exception exp)
        {
            exp.printStackTrace();
            running = false;
        }
        
        if(n != 0) blockSizeStats.StatUpdate(n);
    }

    public void onAdvisory(List<Message> messages, EventQueue eventQueue)
    {
        try
        {
            for (int i = 0,msgNum = messages.size(); i < msgNum; i++)
            {
                System.out.printf("received advisory\n");
                System.out.println(messages.get(i));
            }
        }
        catch(Exception exp)
        {
            // any exception that happens in the callback does not
            // propagate to the caller of the tibEventQueue_Dispatch
            exp.printStackTrace();
        }
    }

    public void recv() throws Exception
    {
        Realm realm   = null;
        TibProperties props   = null;
        EventQueue queue   = null;
        Subscriber subs    = null;
        Subscriber adv     = null;
        ContentMatcher am      = null;

        // Set global trace to specified level
        if (trcLevel != null)
            FTL.setLogLevel(trcLevel);

        // Get a single realm per process
        props = FTL.createProperties();
        props.set(Realm.PROPERTY_STRING_USERNAME, user);
        props.set(Realm.PROPERTY_STRING_USERPASSWORD, password);
        if(identifier!=null)
            props.set(Realm.PROPERTY_STRING_APPINSTANCE_IDENTIFIER, identifier);
        
        if (trustAll || trustFile != null)
        {
            if (realmServer == null)
                realmServer = defaultSecureUrl;

            if (trustAll)
            {
                props.set(Realm.PROPERTY_LONG_TRUST_TYPE, Realm.HTTPS_CONNECTION_TRUST_EVERYONE);
            }
            else
            {
                props.set(Realm.PROPERTY_LONG_TRUST_TYPE, Realm.HTTPS_CONNECTION_USE_SPECIFIED_TRUST_FILE);
                props.set(Realm.PROPERTY_STRING_TRUST_FILE, trustFile);
            }
        }
        else
        {
            if (realmServer == null)
                realmServer = defaultUrl;
        }

        if (clientLabel != null)
            props.set(Realm.PROPERTY_STRING_CLIENT_LABEL, clientLabel);
        else
            props.set(Realm.PROPERTY_STRING_CLIENT_LABEL, this.getClass().getSimpleName());

        realm = FTL.connectToRealmServer(realmServer, appName, props);
        props.destroy();
        
        dataRef = realm.createMessageFieldRef(Message.TIB_BUILTIN_MSG_FMT_OPAQUE_FIELDNAME);
        reportMessage = realm.createMessage("ReportMsgFmt");
        
        props = FTL.createProperties();
        props.set(EventQueue.PROPERTY_BOOL_INLINE_MODE, inline_mode);

        queue = realm.createEventQueue(props);
        props.destroy();

        if (doAdvisory)
        {
            // Create a data loss advisory matcher
            String matcher = "{\""+Advisory.FIELD_NAME+"\":\""+Advisory.NAME_DATALOSS+"\"}";
            
            am = realm.createContentMatcher(matcher);
            // Create an advisory subscriber
            adv = realm.createSubscriber(Advisory.ENDPOINT_NAME, am);

            queue.addSubscriber(adv, new SubscriberListener()
            {
                public void messagesReceived(List<Message> messages, EventQueue eventQueue)
                {
                    onAdvisory(messages,eventQueue);
                }
            });
        }

        props = FTL.createProperties();
        if (durableName != null)
            props.set(Subscriber.PROPERTY_STRING_DURABLE_NAME, durableName);
        subs = realm.createSubscriber(recvEndpointName, null, props);
        props.destroy();
        
        queue.addSubscriber(subs, new SubscriberListener()
        {
            public void messagesReceived(List<Message> messages, EventQueue eventQueue)
            {
                onMsg(messages,eventQueue);
            }
        });

        pub = realm.createPublisher(endpointName);

        while(running)
        {
            try
            {
                queue.dispatch();
            }
            catch(Exception exp)
            {
                exp.printStackTrace();
            }
        }

        pub.close();
        
        if(queue!=null && subs!=null) queue.removeSubscriber(subs);
        if(queue!=null && adv!=null) queue.removeSubscriber(adv);
        
        subs.close();
        if(adv!=null) adv.close();
        if(am!=null) am.destroy();
        queue.destroy();
        dataRef.destroy();
        reportMessage.destroy();
        realm.close();
    }
    
    public static void main(String[] args)
    {
        TibThruRecv s  = new TibThruRecv(args);
        try
        {
            s.recv();
        }
        catch (Throwable e)
        {
            e.printStackTrace();
        }
    } 
}
